feat: Support for training with MoE module in PipelineEngine

shjwudp opened the pull request:
Thanks for the awesome work on MoE! I made some changes to support it running in PipelineEngine

shjwudp commented:
Hi @jeffra @awan-10, would you mind helping to review this PR? Or tell me which of your colleagues I need to request. THANKS :)

awan-10 commented:
Thank you for this PR @shjwudp. Can you please explain if you are trying to support pipeline parallelism and expert parallelism with this PR? Or trying to build an MoE model already built using the PipeEngine? 

@ShadenSmith, FYI.

shjwudp commented:
> Thank you for this PR @shjwudp. Can you please explain if you are trying to support pipeline parallelism and expert parallelism with this PR? Or trying to build an MoE model already built using the PipeEngine?
> 
> @ShadenSmith, FYI.

Hi @awan-10, thank you for your attention! I think I am doing the first function. People can use the MoE module in PipelineEngine, but it doesn't work properly because:
1. The function that creates the expert group does not consider pipeline parallelism. This creates the wrong expert group when using PipeEngine, which causes pipeline parallelism and expert parallelism to not work correctly.
2. MoE checkpoints do not support separate storage according to pipeline stages.  The checkpoint function of PipeEngine can't work correctly.

I fixed these two issues separately to support pipeline parallelism and expert parallelism. BTW, I'm already using this feature in my own branch. This PR is cherry picked. There are some mistake in the middle, which are hard to find. Finally, I added unit tests to ensure correctness.

awan-10 commented:
@shjwudp -- thank you for getting back on this. Do you have a model that you trained using this branch? Can you kindly share the config of that model? 

Jeff and I reviewed some parts of your PR and left some comments/questions for you. 

Also, can you please clarify this point? ```There are some mistake in the middle, which are hard to find``` 
 -- Are you saying there are mistakes in this PR? or mistakes on your branch?

shjwudp commented on tests/unit/test_moe.py:
```diff
@@ -113,3 +109,84 @@ def _test_moe(args, hidden_dim, ep_size, use_residual):
+    def _test_moe(args, hidden_dim):
+        pp_world_size = 2
+        tp_world_size = 1
+        dp_world_size = 2
+        ep_size = 2
+
+        topo = PipeModelDataParallelTopology(num_pp=pp_world_size,
+                                             num_mp=tp_world_size,
+                                             num_dp=dp_world_size)
```
Hi Jeff, thanks for your question!

I think it should not be. For any rank, it should satisfy $ep\\_ size \le dp\\_world\\_size$ and $ep\\_size | dp\\_world\\_size$(expert groups require different data inputs and thus can only be built on dp group ranks).

For the second question, the Expert Parallel Group is currently created with the MoE moudle, which provides good flexibility and supports different MoE moudles with different degrees of expert parallelism (Pyrmaid-MoE relies on this design), I think This is something that Parallel Topology cannot easily do. At the same time, it works fine in both $ep\\_ size \lt dp\\_world\\_size$ and $ep\\_size = dp\\_world\\_size$, so I don't think we need to provide new Parallel Topology.

shjwudp commented on tests/unit/test_moe.py:
```diff
@@ -113,3 +109,84 @@ def _test_moe(args, hidden_dim, ep_size, use_residual):
+    @distributed_test(world_size=[4])
+    def _test_moe(args, hidden_dim):
+        pp_world_size = 2
+        tp_world_size = 1
```
I tried it, but unfortunately it not work. PipeEngine has [limitation](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/pipe/engine.py#L673) that cannot use TP and multi-tensor pipe together. And MoE model transfer hidden states and MoE loss(multi-tensor pipe), this makes TP unable to use.

shjwudp commented:
@awan-10 Yes, I have trained a 2.7B+64E GPT model with the following configuration:

```
--num-layers 32
--hidden-size 2560
--num-attention-heads 16
--global-batch-size 512
--micro-batch-size 4
--tensor-model-parallel-size 1
--pipeline-model-parallel-size 8
--moe-expert-parallel-size 8
--num-experts 64
```

> Are you saying there are mistakes in this PR? or mistakes on your branch?

I believe there is no "mistakes" in this PR.

rocm-mici commented:
Can one of the admins verify this patch?

awan-10 commented:
A lot has changed in terms of how DeepSpeed handles various parallelism strategies.

Sorry for the delay in getting back to this PR @shjwudp. Is it okay to close this PR or are you still actively working on the project?

shjwudp commented:
Hi awan, glad to know you still remember this PR. I haven't paid attention to this issue recently, please close it.
